src/com/amazon/corretto/crypto/provider/HpkeCipher.java

```diff
@@ -0,0 +1,309 @@
+                      inputOffset,
+                      inputLen,
+                      output,
+                      outputOffset));
+      return result;
+    }
+  }
+
+  @Override
+  protected int engineGetOutputSize(int inputLen) {
+    if (params_ == null) {
+      throw new IllegalStateException("params should be set before getOutputSize");
+    }
+    if ((javaCipherMode_ != Cipher.WRAP_MODE) && (javaCipherMode_ != Cipher.UNWRAP_MODE)) {
+      throw new IllegalStateException("cipher mode should be set before getOutputSize");
+    }
+    return hpkeOutputSize(
+        javaCipherMode_, params_.getKemId(), params_.getKdfId(), params_.getAeadId(), inputLen);
+  }
+
+  // Boilerplate Methods
+  // -------------------
+
+  @Override
+  protected AlgorithmParameters engineGetParameters() {
+    AlgorithmParameters params;
+    try {
+      params = AlgorithmParameters.getInstance("HPKE");
+      params.init(params_);
+      return params;
+    } catch (NoSuchAlgorithmException | InvalidParameterSpecException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected byte[] engineDoFinal(byte[] input, int inputOffset, int inputLen)
+      throws IllegalBlockSizeException, BadPaddingException {
+    byte[] output = new byte[engineGetOutputSize(inputLen)];
+    try {
+      int len = engineDoFinal(input, inputOffset, inputLen, output, 0);
+      if (len != output.length) {
+        throw new RuntimeCryptoException(
+            "HpkeCipher expected output of length " + output.length + ", got output of len " + len);
+      }
+      return output;
+    } catch (ShortBufferException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected void engineInit(int opmode, Key key, AlgorithmParameters params, SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if (params == null) {
+      throw new InvalidAlgorithmParameterException("cannot initialize HpkeCipher with null params");
+    }
+    try {
+      AlgorithmParameterSpec spec = params.getParameterSpec(HpkeParameterSpec.class);
+      engineInit(opmode, key, spec, random);
+    } catch (InvalidParameterSpecException e) {
+      throw new InvalidAlgorithmParameterException(e);
+    }
+  }
+
+  @Override
+  protected void engineInit(int opmode, Key key, SecureRandom random) throws InvalidKeyException {
+    if (key instanceof EvpHpkeKey) {
+      try {
+        engineInit(opmode, key, ((EvpHpkeKey) key).getSpec(), random);
+      } catch (InvalidAlgorithmParameterException e) {
+        throw new InvalidKeyException(e);
+      }
+    } else {
+      throw new InvalidKeyException("HpkeCipher can only be initialized with EvpHpkeKey.");
+    }
+  }
+
+  // Unsupported Methods
+  // -------------------
+
+  @Override
+  protected void engineUpdateAAD(byte[] src, int offset, int len) {
+    // TODO: implement AAD support
+    throw new IllegalStateException("HpkeCipher currently does not support AAD");
```

Comment:
This limitation is a bit surprising since all the parameter sets we're supporting have an AEAD cipher. Is this because we ran out of time or is there a another reason for this limitation?

Reply:
I ran out of time and wanted to open a PR to get feedback on the general structure. I will fix this in a subsequent commit (before the PR is merged.)

Reply:
I added AAD support